### DISCUSSION TITLE
How can I get a list of alerts using the API?

### DISCUSSION THREAD

**Discussion by danfinn** [grafana/grafana#60657] [Alerting]
I'm finding the grafana documentation really lacking around the alerting API, also quite confusing.  It's not clear which docs I should be using, there are multiple alerting related APIs referenced and it's not clear which is correct.

I'm trying to migrate my legacy setup of old style alerts to the new way of doing things, which I think is called "unified alerting".   What I'd like to be able to do is configure my alerts using the UI, export those via API (or UI if that's possible) and then load those via a file from git (we do this now for dashboards and other config).  Is this possible?  If so, could someone help me with the correct API to use?

Through some googling I just found this, which I have not seen mentioned anywhere in the grafana docs but it does seem to at least return the alerts that are configured:
https://community.grafana.com/t/grafana-new-unified-alerting-api/70156/3

Is this the correct one to use and if so, why is this not mentioned in the grafana docs?  Instead I see references to the Alerting provisioning API which does not seem to have a way to dump all of the alerts from what I can tell.

**Comment by danfinn** [comment]
An example of the confusion.  This link, which is for the latest version, refers on step 2 to the deprecated alerting api:

https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/file-provisioning/

Towards the top of the URL for the deprecated API (https://grafana.com/docs/grafana/latest/developers/http_api/alerting/#get-alerts) it says "If you are using Grafana Alerting...".  How do I know if I am or not, this is not clear.  Assuming that this is what I'm trying to use, that takes me to:

https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/

The alerting provisioning API which does not seem to have the ability (or is not documented anyways) to dump all of the currently configured alerts.  It seems that you have to know a specific alert ID (where do I get that from?) to be able to query the API.

**Comment by danfinn** [comment]
More confusion, I found this link:
https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/

Which gives you essentially no information on how to provision from a file other than telling you it's a possibility however at the bottom of that page are 3 "Useful Links", the first of which is Grafana Provisioning which takes you to:

https://grafana.com/docs/grafana/latest/administration/provisioning/

In the Alerting section of this page (https://grafana.com/docs/grafana/latest/administration/provisioning/#alerting)  it says:
"For information on provisioning Grafana Alerting, refer to [Provision Grafana Alerting resources](https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/)."

Taking you back full circle to the first link.  There seems to be quite a bit of information lacking here and much room for improvement.

**Comment by gelicia** [comment]
I think one confusing thing is this page doesn't seem to be linked the same way other pages are https://grafana.com/docs/grafana/latest/developers/http_api/alerting/ 

This defines legacy alerting vs new alerting. The new API is the alerting provisioning one which is probably what you want to use. Looking at that, theres the `/v1/provisioning/alert-rules/` which says the UID is required but is not when I query it locally. 

I can log a bug ticket to the docs team on the alerting documentation stuff, but is that what you're wanting to do ?

**Comment by danfinn** [comment]
Unfortunately that is not working for me.  If I query http://localhost:3000/api/v1/provisioning/alert-rules
I get:
 {"message":"Not found"}.

If I try an arbitrary alert ID, like http://localhost:3000/api/v1/provisioning/alert-rules/4 I get
{"message":"could not find alert rule","traceID":""}

The API key I'm using is an Admin role.

Regarding the ticket for the documentation, I most certainly think there are major issues with the documentation around provisioning alerts and the related APIs and that it should be fixed/improved.

**Reply by FrankAnk** [reply]
I'm having the same problem. There just doesn't seem to be an api to retrieve all alert-rules.

**Comment by vxe** [comment]
this seems to get some of them: https://editor.swagger.io/?url=https://raw.githubusercontent.com/grafana/grafana/main/pkg/services/ngalert/api/tooling/post.json , but you need to hit an alertmanager endpoint so idk if this is the general solution

**Comment by MandaloreLC** [comment]
I think you need the `/api/v1/provisioning/alert-rules` endpoint.
Please refer to https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/

**Reply by FrankAnk** [reply]
Yes, that is what we all would expect.
The Problem is that the specified URL does not return any alerts. I only get a 404 with body  {"message": "Not found"}

Grafana v9.3.6

**Reply by FrankAnk** [reply]
I have just upgraded Grafana from 9.3.6 to 9.5.1 and the problem seems to have been fixed somewhere between these versions.
The endpoint ```/api/v1/provisioning/alert-rules``` now returns the expected alert-rules.

**Reply by konrad147** [reply]
The answer is correct. The `/api/v1/provisioning/alert-rules` has been added to Grafana in version 9.4

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
